### PR TITLE
Support Iceberg positional deletes

### PR DESCRIPTION
**Pull request by yingsu00**
This PR is the first cut to support reading Iceberg tables with positional delete files. 
Design doc: https://github.com/facebookincubator/velox/issues/5977

TODO:
- Delete file deserialization in HiveIcebergSplit. This will be done after the Prestissimo changes are in shape
- File path matching when reading the positional delete files
- Performance optimization to avoid extra conversion from row numbers to bitmap then to row numbers

Catches:
- The velox_hive_iceberg_datasource would be better to be included in velox_hive_connector. But so far it's unclear to me how to do it properly. Suggestions are welcome.

### CODE REVIEW

File: velox/connectors/hive/iceberg/IcebergDataSource.cpp [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+  // TODO: Deserialize the std::vector<IcebergDeleteFile> deleteFiles. For now
+  // we assume it's already deserialized.
+  std::shared_ptr<HiveIcebergSplit> icebergSplit =
+      std::dynamic_pointer_cast<HiveIcebergSplit>(split);
+
+  auto& deleteFiles = icebergSplit->deleteFiles;
+  for (auto& deleteFile : deleteFiles) {
+    if (deleteFile.content == FileContent::kEqualityDeletes) {
+      readEqualityDeletes(deleteFile);
```
Comment (majetideepak) [inline velox/connectors/hive/iceberg/IcebergDataSource.cpp:76]: Move equality deletes to another PR.

File: velox/connectors/hive/storage_adapters/gcs/examples/CMakeLists.txt [hunk 2]
```diff
@@ -20,5 +20,6 @@ target_link_libraries(
   velox_gcs
   velox_core
   velox_hive_connector
+  velox_hive_iceberg_datasource
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/examples/CMakeLists.txt:23]: These CMakeLists.txt additions should be redundant since `velox_hive_iceberg_datasource` is added to `velox_hive_connector`.

**Review by majetideepak — COMMENTED** [review]
@yingsu00 Some high-level comments while we converge on the overall design.

**Review by Yuhta — COMMENTED** [review]
@yingsu00 We probably want to create some `DeltaProcessor` interface class like this and call it in different parts of `HiveDataSource`, instead of subclassing `HiveDataSource`, because the split is not guaranteed to be the same type during the life cycle of the data source.  At end of the day, `DeltaProcessor` is the property of split, not of data source.

```c++
class DeltaProcessor {
  static std::unique_ptr<DeltaProcessor> create(const HiveConnectorSplit&);

  // Called in HiveDataSource::addSplit to prepare delta readers.
  void prepareSplit();

  void readNext(RowReader&, int64_t size, VectorPtr& output);

  // In case delta readers need to be aware of filters.
  void addDynamicFilter();

  // Flush the data in delta reader out (e.g., in case we split update into delete + append, the newly appended rows need to be flushed here).
  void finishReadSplit();
};
```

Let me know how do you think.  I can create some skeleton about this class and implement it for basic case (no delta).

File: velox/connectors/hive/iceberg/IcebergDataSource.cpp [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+  openDeletes(deleteFile, deleteRowType, deleteColumnHandles);
+}
+
+void HiveIcebergDataSource::readEqualityDeletes(
+    const IcebergDeleteFile& deleteFile) {
+  VELOX_NYI();
+}
+
+void HiveIcebergDataSource::openDeletes(
```
[the remark this answers is not in the document]
Comment (yingsu00) [reply velox/connectors/hive/iceberg/IcebergDataSource.cpp:167]: @Yuhta I intended to do this in future PRs. Do we have to do it now?

File: velox/connectors/hive/storage_adapters/gcs/examples/CMakeLists.txt (hunk 2 above)
Comment (yingsu00) [reply velox/connectors/hive/storage_adapters/gcs/examples/CMakeLists.txt:23]: @majetideepak It didn't work and would throw link errors. Do you know how to make velox_hive_connector contain the velox_hive_iceberg_datasource? I tried to add PUBLIC link lib to velox_hive_iceberg_datasource, but it also doesn't work.

File: velox/connectors/hive/iceberg/IcebergSplit.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    const std::shared_ptr<std::string>& _extraFileInfo)
+    : HiveConnectorSplit(
+          _connectorId,
+          _filePath,
+          _fileFormat,
+          _start,
+          _length,
+          _partitionKeys,
+          _tableBucketNumber) {
+  // TODO: Deserialize _extraFileInfo to get deleteFiles;
```
[the remark this answers is not in the document]
Comment (yingsu00) [reply velox/connectors/hive/iceberg/IcebergSplit.cpp:40]: > finish TODO.

The delete file serialization is not done on the Presto_cpp yet, so I plan to do this in future PR after we decide how to serialize the delete files.

**Comment by yingsu00** [comment]
>  because the split is not guaranteed to be the same type during the life cycle of the data source

@Yuhta Thanks for brainstorming. Could you give me an example when this is going to happen?

File: velox/connectors/hive/iceberg/IcebergDataSource.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yuhta) [reply velox/connectors/hive/iceberg/IcebergDataSource.cpp:167]: This will have a rather big impact on how you structure the thing, so better to do it right now, otherwise most of what you write now need to be rewritten.

**Comment by Yuhta** [comment]
@yingsu00 For now this is not switch between two different frameworks, but could happen that the first split is a normal split without deltas but the subsequent one is a split with deltas.  So process it at per split level looks the right thing to me.

**Comment by yingsu00** [comment]
> @yingsu00 For now this is not switch between two different frameworks, but could happen that the first split is a normal split without deltas but the subsequent one is a split with deltas. So process it at per split level looks the right thing to me.

This is fine. But the splits for reading an Iceberg table will always come in with extraFileInfo["table_format"]="iceberg"

**Comment by Yuhta** [comment]
> This is fine. But the splits for reading an Iceberg table will always come in with extraFileInfo["table_format"]="iceberg"

This is not guaranteed for future table formats

**Comment by yingsu00** [comment]
> > This is fine. But the splits for reading an Iceberg table will always come in with extraFileInfo["table_format"]="iceberg"
> 
> This is not guaranteed for future table formats

But Iceberg splits will always come in with this, similar to Hudi. We can guarantee that and their datasource can be constructed accordingly.

**Comment by Yuhta** [comment]
> > > This is fine. But the splits for reading an Iceberg table will always come in with extraFileInfo["table_format"]="iceberg"
> > 
> > 
> > This is not guaranteed for future table formats
> 
> But Iceberg splits will always come in with this, similar to Hudi. We can guarantee that and their datasource can be constructed accordingly.

@yingsu00 We are design the framework for all these table formats, so dispatch at split level is the right thing here

**Comment by yingsu00** [comment]
> > > > This is fine. But the splits for reading an Iceberg table will always come in with extraFileInfo["table_format"]="iceberg"
> > > 
> > > 
> > > This is not guaranteed for future table formats
> > 
> > 
> > But Iceberg splits will always come in with this, similar to Hudi. We can guarantee that and their datasource can be constructed accordingly.
> 
> @yingsu00 We are design the framework for all these table formats, so dispatch at split level is the right thing here

But you can't mix everything together. These table formats spec are very different to each other and they are quite complex. THere has to be some abstractions, layers and separation of concerns. And how was the decision made?

**Comment by assignUser** [comment]
Superseded by #7362
